File: be2/app/controller/faces_controller.py
import json
import math
from models import faces as model_face

def calculate_euclidean_distance(a, b):
    return math.sqrt(sum((x - y) ** 2 for x, y in zip(a, b)))

def find_nearest(user_embedding, db_embeddings):
    best_match = None
    smallest_distance = float('inf')

    for row in db_embeddings:
        embedding_str = row.get("embedding")
        if not embedding_str:
            continue
        known_embedding = list(map(float, embedding_str.split(",")))
        if len(known_embedding) != len(user_embedding):
            continue

        distance = calculate_euclidean_distance(user_embedding, known_embedding)
        if distance < smallest_distance:
            smallest_distance = distance
            best_match = {
                "name": row.get("name"),
                "distance": distance
            }

    return best_match

async def verify_face_handler(websocket, msg: dict):
    user_embedding = msg.get("embedding")
    print("Message received:", msg)
    print("Embedding:", user_embedding)

    if not isinstance(user_embedding, list):
        await websocket.send_text(json.dumps({
            "type": "error",
            "message": "Invalid or missing 'embedding'"
        }
    )
)
        return

    try:
        results = await model_face.get_all_faces()
        match = find_nearest(user_embedding, results)

        if match:
            await websocket.send_text(json.dumps({
                "type": "recognize_face",
                "match": match["distance"] < 0.9,
                "name": match["name"],
                "distance": match["distance"]
            }
        )
    )
        else:
            await websocket.send_text(json.dumps({
                "type": "recognize_face",
                "match": False,
                "name": None,
                "distance": None
            }
        )
    )
    except Exception as e:
        print("Verify face error:", e)
        await websocket.send_text(json.dumps({
            "type": "error",
            "message": "Server error during recognition"
        }
    )
)

async def insert_face_handler(websocket, msg: dict):
    try:
        name = msg.get("name")
        embedding = msg.get("embedding")
        print("Message received:", msg)

        if not name or not isinstance(embedding, list):
            await websocket.send_text(json.dumps({
                "type": "insert_face",
                "success": False,
                "message": "Invalid name or embedding format. 'embedding' must be an array."
            }
        )
    )
            return

        embedding_string = ",".join(map(str, embedding))
        await model_face.insert_face(name, embedding_string)

        await websocket.send_text(json.dumps({
            "type": "insert_face",
            "success": True
        }
    )
)
    except Exception as e:
        print("Insert face error:", e)
        await websocket.send_text(json.dumps({
            "type": "insert_face",
            "success": False,
            "message": str(e)
        }
    )
)

File: be3/src/controller/facesController.js
const modelFace = require("../models/faces");

function calculateEuclideanDistance(a, b) {
    let sum = 0;
    for (let i = 0; i < a.length; i++) {
        const diff = a[i] - b[i];
        sum += diff * diff;
    }
    return Math.sqrt(sum);
}

// Find nearest embedding in DB
function findNearest(userEmbedding, dbEmbeddings) {
    let bestMatch = null;
    let smallestDistance = Infinity;

    for (let row of dbEmbeddings) {
        const knownEmbedding = row.embeddings?.split(",").map(Number);
        if (!knownEmbedding || knownEmbedding.length !== userEmbedding.length) continue;

        const distance = calculateEuclideanDistance(userEmbedding, knownEmbedding);
        if (distance < smallestDistance) {
            smallestDistance = distance;
            bestMatch = {
                name: row.name,
                distance
            };
        }
    }

    return bestMatch;
}

// VERIFY FACE HANDLER
const verifyFaceHandler = async (ws, msg) => {
    const userEmbedding = msg.embeddings;
    console.log("Message received:", msg);
    console.log("Embedding:", msg.embeddings);
    if (!userEmbedding || !Array.isArray(userEmbedding)) {
        ws.send(JSON.stringify({
            type: "error",
            message: "Invalid or missing 'embedding'"
        }));
        return;
    }

    try {
        const [results] = await modelFace.getUserFaceData();
        const match = findNearest(userEmbedding, results);

        if (match) {
            ws.send(JSON.stringify({
                type: "recognize_face",
                match: match.distance < 0.9,
                name: match.name,
                distance: match.distance
            }));
        } else {
            ws.send(JSON.stringify({
                type: "recognize_face",
                match: false,
                name: null,
                distance: null
            }));
        }
    } catch (error) {
        console.error("Verify face error:", error);
        ws.send(JSON.stringify({
            type: "error",
            message: "Server error during recognition"
        }));
    }
};

// INSERT FACE HANDLER
const insertFaceHandler = async (ws, msg) => {
    try {
        const { name, email, phone, embeddings } = msg;
        console.log("Message received:", msg);
        console.log("Name:", msg.name);
        console.log("Email:", msg.email);
        console.log("Phone Number:", msg.phone);
        console.log("Embedding:", msg.embeddings);
        if (!name || !email || !phone || !embeddings || !Array.isArray(embeddings)) {
            return ws.send(JSON.stringify({
                type: "insert_face",
                success: false,
                message: "Invalid name or embedding format. 'embedding' must be an array."
            }));
        }

        const embeddingString = embeddings.join(",");
        await modelFace.insertUsers(name, email, phone, embeddingString);

        ws.send(
            JSON.stringify({
                type: "insert_face",
                success: true,
            })
        );
    } catch (error) {
        console.error("Insert face error:", error);
        ws.send(
            JSON.stringify({
                type: "insert_face",
                success: false,
                message: error.message,
            })
        );
    }
};



module.exports = {
    verifyFaceHandler,
    insertFaceHandler
}